sharding/contracts/sharding_manager_test.go

```diff
@@ -1,130 +1,560 @@
+
+	notary, err = smc.NotaryRegistry(&bind.CallOpts{}, notaryPoolAddr[0])
+
+	if !notary.Deposited &&
+		notary.PoolIndex.Cmp(big.NewInt(0)) != 0 &&
+		notary.DeregisteredPeriod.Cmp(big.NewInt(0)) != 0 {
+		t.Fatalf("Incorrect notary registry. Want - deposited:true, index:0, period:0"+
+			"Got - deposited:%v, index:%v, period:%v ", notary.Deposited, notary.PoolIndex, notary.DeregisteredPeriod)
+	}
+
+	// Test notary 1 and 2 have registered
+	if _, err := smc.RegisterNotary(txOpts[1]); err != nil {
+		t.Fatalf("Registering notary has failed: %v", err)
+	}
+	backend.Commit()
+
+	if _, err := smc.RegisterNotary(txOpts[2]); err != nil {
+		t.Fatalf("Registering notary has failed: %v", err)
+	}
+	backend.Commit()
+
+	notary, err = smc.NotaryRegistry(&bind.CallOpts{}, notaryPoolAddr[1])
+
+	if !notary.Deposited &&
+		notary.PoolIndex.Cmp(big.NewInt(1)) != 0 &&
+		notary.DeregisteredPeriod.Cmp(big.NewInt(0)) != 0 {
+		t.Fatalf("Incorrect notary registry. Want - deposited:true, index:1, period:0"+
+			"Got - deposited:%v, index:%v, period:%v ", notary.Deposited, notary.PoolIndex, notary.DeregisteredPeriod)
+	}
+
+	notary, err = smc.NotaryRegistry(&bind.CallOpts{}, notaryPoolAddr[2])
+
+	if !notary.Deposited &&
+		notary.PoolIndex.Cmp(big.NewInt(2)) != 0 &&
+		notary.DeregisteredPeriod.Cmp(big.NewInt(0)) != 0 {
+		t.Fatalf("Incorrect notary registry. Want - deposited:true, index:2, period:0"+
+			"Got - deposited:%v, index:%v, period:%v ", notary.Deposited, notary.PoolIndex, notary.DeregisteredPeriod)
+	}
+
+	// Check total numbers of notaries in pool
+	numNotaries, err := smc.NotaryPoolLength(&bind.CallOpts{})
 	if err != nil {
-		t.Fatalf("Error getting collationGasLimit: %v", err)
+		t.Fatalf("Failed to get notary pool length: %v", err)
 	}
-	if gasLimit.Cmp(big.NewInt(10000000)) != 0 {
-		t.Fatalf("collation gas limit should be 10000000 gas")
+	if numNotaries.Cmp(big.NewInt(3)) != 0 {
+		t.Fatalf("Incorrect count from notary pool. Want: 3, Got: %v", numNotaries)
 	}
 }
 
-// Test collator deposit
-func TestCollatorDeposit(t *testing.T) {
-	backend := backends.NewSimulatedBackend(core.GenesisAlloc{addr: {Balance: accountBalance1001Eth}})
-	transactOpts := bind.NewKeyedTransactor(key)
-	_, _, smc, _ := deploySMCContract(backend)
+func TestNotaryRegisterInsufficientEther(t *testing.T) {
+	addr := crypto.PubkeyToAddress(mainKey.PublicKey)
+	backend := backends.NewSimulatedBackend(core.GenesisAlloc{addr: {Balance: accountBalance2000Eth}})
+	txOpts := bind.NewKeyedTransactor(mainKey)
+	txOpts.Value = notaryDepositInsufficient
+	_, _, smc, _ := deploySMCContract(backend, mainKey)
+
+	smc.RegisterNotary(txOpts)
+	backend.Commit()
+
+	notary, _ := smc.NotaryRegistry(&bind.CallOpts{}, addr)
+	numNotaries, _ := smc.NotaryPoolLength(&bind.CallOpts{})
+
+	if notary.Deposited {
+		t.Fatalf("Notary deposited with insufficient fund")
+	}
+
+	if numNotaries.Cmp(big.NewInt(0)) != 0 {
+		t.Fatalf("Incorrect count from notary pool. Want: 0, Got: %v", numNotaries)
+	}
```

Comment:
I think there's a bit more of the story to tell here. Does the transaction succeed? Does the sender lose all of the ETH they sent?

Reply:
The sender gets the ETH back, but all the gas gets consumed, there's no gas refund